File: python/sort/selection_sort.py


def selection_sort(arr: list) -> list:
    for i in range(len(arr)):
        min_index = i
        for j in range(i+1, len(arr)):
            if arr[j] < arr[min_index]:
                min_index = j

        if i != min_index:
            print(i, min_index)
            arr[i], arr[min_index] = arr[min_index], arr[i]
            print("swapped happen")

    return arr


a = [1, 2, 100, 3, 99, 200, 4, 5]
print(selection_sort(a))

File: javascript/sort/selectionSort.js
const swap = (arr, i1, i2) => ([arr[i1], arr[i2]] = [arr[i2], arr[i1]]);

/**
 *
 * @param {Array} arr
 */
function selectionSort(arr) {
  for (let i = 0; i < arr.length; i++) {
    let minIndex = i;
    for (let j = i + 1; j < arr.length; j++) {
      // console.log(arr, arr[j], arr[minIndex]);
      // console.log(i, j);
      if (arr[j] < arr[minIndex]) {
        minIndex = j;
      }
    }
    if (i !== minIndex) {
      console.log("swap happened");
      console.log(
        `i=${i} minIndex=${minIndex} arr[i]=${arr[i]} arr[minIndex]=${arr[minIndex]}`
      );
      swap(arr, minIndex, i);
    }
  }
  return arr;
}

let a = [1, 2, 100, 3, 99, 200, 4, 5];
console.log(selectionSort(a));
